test/threadstest.c

```diff
@@ -741,6 +745,330 @@ static int test_bio_dgram_pair(void)
+static uint32_t test_thread_native_fn(void *data)
+{
+    uint32_t *ldata = (uint32_t*) data;
+    *ldata = *ldata + 1;
+    return *ldata - 1;
+}
+
+static uint32_t test_thread_noreturn(void *data)
+{
+    CRYPTO_MUTEX *lock = (uint32_t*) data;
+
+    /* lock is assumed to be locked */
+    ossl_crypto_mutex_lock(lock);
+
+    /*
+     * if we ever get here, the test is written wrong; to ensure
+     * that this function is ossl_noreturn, call OPENSSL_die
+     */
+    OPENSSL_die("test_thread_noreturn", __FILE__, __LINE__);
+    return 0;
+}
+
+/* Tests of native threads */
+
+static int test_thread_native(void)
+{
+    int testval = 0;
+    uint32_t retval;
+    uint32_t local;
+    CRYPTO_THREAD *t;
+    CRYPTO_MUTEX *lock;
+
+    /* thread spawn, join and termination */
+
+    local = 1;
+    t = ossl_crypto_thread_native_start(test_thread_native_fn, &local, 1);
+    if (!TEST_ptr(t))
+        return 0;
+
+    /*
+     * pthread_join results in undefined behaviour if called on a joined
+     * thread. We do not impose such restrictions, so it's up to us to
+     * ensure that this does not happen (thread sanitizer will warn us
+     * if we do).
+     */
+    if (!TEST_int_eq(ossl_crypto_thread_native_join(t, &retval), 1))
+        return 0;
+    if (!TEST_int_eq(ossl_crypto_thread_native_join(t, &retval), 1))
+        return 0;
+
+    if (!TEST_int_eq(retval, 1) || !TEST_int_eq(local, 2))
+        return 0;
+
+    if (!TEST_int_eq(ossl_crypto_thread_native_terminate(t), 1))
+        return 0;
+    if (!TEST_int_eq(ossl_crypto_thread_native_terminate(t), 1))
+        return 0;
+
+    if (!TEST_int_eq(ossl_crypto_thread_native_join(t, &retval), 1))
+        return 0;
+
+    if (!TEST_int_eq(ossl_crypto_thread_native_clean(t), 1))
+        return 0;
+    t = NULL;
+
+    if (!TEST_int_eq(ossl_crypto_thread_native_clean(t), 0))
+        return 0;
+
+    /* termination of a long running thread */
+
+    lock = ossl_crypto_mutex_new();
+    if (!TEST_ptr(lock))
+        return 0;
+    ossl_crypto_mutex_lock(lock);
+
+    t = ossl_crypto_thread_native_start(test_thread_noreturn, lock, 1);
+    if (!TEST_ptr(t))
+        goto fail;
+    if (!TEST_int_eq(ossl_crypto_thread_native_terminate(t), 1))
+        goto fail;
+    if (!TEST_int_eq(ossl_crypto_thread_native_clean(t), 1))
+        goto fail;
+
+    testval = 1;
+
+fail:
+    ossl_crypto_mutex_unlock(lock);
+    ossl_crypto_mutex_free(&lock);
+    if (!TEST_ptr_null(lock))
+        return 0;
+
+    return testval;
+}
+
+#if !defined(OPENSSL_NO_DEFAULT_THREAD_POOL)
+static int test_thread_internal(void)
+{
+    uint32_t retval[3];
+    uint32_t local[3] = { 0 };
+    uint32_t threads_supported;
+    size_t i;
+    void *t[3];
+    OSSL_LIB_CTX *cust_ctx = OSSL_LIB_CTX_new();
+
+    threads_supported = OSSL_get_thread_support_flags();
+    threads_supported &= OSSL_THREAD_SUPPORT_FLAG_DEFAULT_SPAWN;
+
+    /* currently only the default pool is supported */
```

Comment:
Is this accurate? I don't see how this restriction occurs in the code, can you point me to it?

If this is true it should be documented in the man page.